pkg/apis/configuration/validation/validation.go

```diff
@@ -596,15 +596,140 @@ func validateRoute(route v1.Route, fieldPath *field.Path, upstreamNames sets.Str
 func validateAction(action *v1.Action, fieldPath *field.Path, upstreamNames sets.String) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if action.Pass == "" {
-		return append(allErrs, field.Required(fieldPath.Child("pass"), ""))
+	if action.Pass == "" && action.Redirect == nil {
+		return append(allErrs, field.Required(fieldPath, "action must specify exactly one of `pass` or `redirect`"))
 	}
 
-	allErrs = append(allErrs, validateReferencedUpstream(action.Pass, fieldPath.Child("pass"), upstreamNames)...)
+	if action.Pass != "" && action.Redirect != nil {
+		return append(allErrs, field.Required(fieldPath, "action must specify exactly one of `pass` or `redirect`"))
+	}
+
+	if action.Pass != "" {
+		allErrs = append(allErrs, validateReferencedUpstream(action.Pass, fieldPath.Child("pass"), upstreamNames)...)
+	}
+
+	if action.Redirect != nil {
+		allErrs = append(allErrs, validateActionRedirect(action.Redirect, fieldPath.Child("redirect"))...)
+	}
 
 	return allErrs
 }
 
+func validateActionRedirect(redirect *v1.ActionRedirect, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, validateRedirectURL(redirect.URL, fieldPath.Child("url"))...)
+
+	if redirect.Code != 0 {
+		allErrs = append(allErrs, validateRedirectStatusCode(redirect.Code, fieldPath.Child("code"))...)
+	}
+
+	return allErrs
+}
+
+// Returns a slice of vars enclosed in ${}. For example ${scheme} would return scheme.
+func captureVariables(s string) [][]string {
+	regex := regexp.MustCompile(`\$\{([^}]+)\}`)
```

Comment:
I was thinking, maybe we can have this regular expression as a global var outside of the method for consistency? I'm ok with having it here too tho.